geomstats/stratified_geometry/spider.py

```diff
@@ -0,0 +1,303 @@
+
+    Parameters
+    ----------
+    s : int
+        The stratum, an integer indicating the stratum the point lies in. If zero, then
+        the point is on the origin.
+    x : float
+        A positive number, the coordinate of the point. It must be zero if and only if
+        the stratum is zero, i.e. the origin.
+    """
+
+    def __init__(self, s, x):
+        super(SpiderPoint, self).__init__()
+        if s == 0 and x != 0:
+            raise ValueError("If the stratum is zero, x must be zero.")
+        self.s = s
+        self.x = x
+
+    def __repr__(self):
+        """Return a readable representation of the instance."""
+        return f"s{self.s}: {self.x}"
+
+    def __hash__(self):
+        """Return the hash of the instance."""
+        return hash((self.s, self.x))
+
+    def __eq__(self, other):
+        """Compare two points."""
+        return self.s == other.s and abs(self.x - other.x) < gs.atol
+
+    def to_array(self):
+        """Return the hash of the instance."""
+        return gs.array([self.s, self.x])
+
+
+class Spider(PointSet):
+    r"""Spider: a set of rays attached to the origin.
+
+    Parameters
+    ----------
+    rays : int
+        Number of rays to attach to the origin.
+        Note that zero counts as the origin not as a ray.
+    """
+
+    def __init__(self, rays):
+        super(Spider, self).__init__()
+        self.rays = rays
+
+    def random_point(self, n_samples=1):
+        r"""Compute a random point of the spider set.
+
+        Parameters
+        ----------
+        n_samples : int
+            Number of samples.
+            Optional, default: 1.
+
+        Returns
+        -------
+        samples : list of SpiderPoint, shape=[...]
+            List of SpiderPoints randomly sampled from the Spider.
+        """
+        if self.rays != 0:
+            s = gs.random.randint(low=0, high=self.rays, size=n_samples)
+            x = gs.abs(gs.random.normal(loc=10, scale=1, size=n_samples))
+            x[s == 0] = 0
+            return [SpiderPoint(s=s[k], x=x[k]) for k in range(n_samples)]
+        return [SpiderPoint(s=0, x=0)] * n_samples
+
+    @_vectorize_point((1, "point"))
+    def belongs(self, point):
+        r"""Check if a random point belongs to the spider set.
+
+        Parameters
+        ----------
+        point : SpiderPoint or list of SpiderPoint, shape=[...]
+             Point to be checked.
+
+        Returns
+        -------
+        belongs : array-like, shape=[...]
+            Boolean denoting if the SpiderPoint belongs to the Spider Set.
+        """
+        results = []
+        for single_point in point:
+            results += [
+                self._value_check(single_point)
+                and self._rays_check(single_point)
+                and self._zero_check(single_point)
+                and type(single_point) is SpiderPoint
+            ]
+        return gs.array(results)
+
+    def _rays_check(self, single_point):
```

Comment:
n_rays

Reply:
I was wondering what was `single_point`, but I understand that you have to make the distinction between `single_point` and `point`. The single_points are arranged into lists, and not into an array.

Reply:
Exactly. Multiple points ara arrenged into lists. We built a decorator which allows the functions to work for both a single point and a list of points, to mantain the same "vectorized" fashion of the rest of the package.